File: lab3/zad1/cpp/stack.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include "node.h"
#include "stack.h"

using namespace std;

Stack::Stack() {
    top = nullptr;
}

Stack::~Stack() {
    while (top != nullptr) {
        StackNode* temp = top;
        top = top->next;
        delete temp;
    }
}

void Stack::push(const string& val) {
    auto* newNode = new StackNode(val);
    newNode->next = top;
    top = newNode;
}

void Stack::pop() {
    if (top == nullptr) {
        throw runtime_error("error");
    }
    StackNode* temp = top;
    top = top->next;
    delete temp;
}

void Stack::print() const {
    if (top == nullptr) {
        cout << "error" << '\n';
        return;
    }
    
    StackNode* curr = top;
    cout << "стек: ";
    while (curr != nullptr) {
        cout << curr->data;
        if (curr->next != nullptr) {
            cout << " -> ";
        }
        curr = curr->next;
    }
    cout << '\n';
}

void Stack::saveToFile(ofstream& file) const {
    if (!file.is_open()) {
        return;
    }
    StackNode* curr = top;
    while (curr != nullptr) {
        file << curr->data << " ";
        curr = curr->next;
    }
}

void Stack::loadFromFile(ifstream& file) {
    if (!file.is_open()) {
        return;
    }
    vector<string> elements;
    string value;
    while (file >> value) {
        elements.push_back(value);
    }

    for (auto it = elements.rbegin(); it != elements.rend(); ++it) {
        push(*it);
    }
}

void Stack::saveToBinaryFile(ofstream& file) const {
    if (!file.is_open()) {
        return;
    }
    StackNode* curr = top;
    while (curr != nullptr) {
        size_t strSize = curr->data.size();
        auto streamSize = static_cast<streamsize>(strSize);
        file.write(reinterpret_cast<const char*>(&strSize), sizeof(strSize));
        file.write(curr->data.c_str(), streamSize);
        curr = curr->next;
    }
}

void Stack::loadFromBinaryFile(ifstream& file) {
    if (!file.is_open()) {
        return;
    }
    vector<string> elements;
    
    while (file.peek() != EOF) {
        size_t strSize;
        file.read(reinterpret_cast<char*>(&strSize), sizeof(strSize));
        if (file.eof()){ 
            break;
        }
        auto streamSize = static_cast<streamsize>(strSize);
        char* buffer = new char[strSize + 1];
        file.read(buffer, streamSize);
        buffer[strSize] = '\0'; 
        
        elements.emplace_back(buffer);
        delete[] buffer;
    }
    
    for (auto it = elements.rbegin(); it != elements.rend(); ++it) {
        push(*it);
    }
}

File: lab3/zad1/go/stack/stack.go
package stack

import (
	"encoding/binary"
	"errors"
	"fmt"
	"io"
	"os"
	"strings"
	"goproj/node"
)

type Stack struct {
	top *node.StackNode
}

func NewStack() *Stack {
	return &Stack{}
}

func (s *Stack) Push(val string) {
	newNode := &node.StackNode{Data: val}
	newNode.Next = s.top
	s.top = newNode
}

func (s *Stack) Pop() error {
	if s.top == nil {
		return errors.New("error")
	}
	s.top = s.top.Next
	return nil
}

func (s *Stack) Print() {
	if s.top == nil {
		fmt.Println("error")
		return
	}

	curr := s.top
	fmt.Print("стек: ")
	for curr != nil {
		fmt.Print(curr.Data)
		if curr.Next != nil {
			fmt.Print(" -> ")
		}
		curr = curr.Next
	}
	fmt.Println()
}

func (s *Stack) SaveToFile(file *os.File) error {
	if file == nil {
		return errors.New("file is nil")
	}

	curr := s.top
	for curr != nil {
		_, err := file.WriteString(curr.Data + " ")
		if err != nil {
			return err
		}
		curr = curr.Next
	}
	return nil
}

func (s *Stack) LoadFromFile(file *os.File) error {
	if file == nil {
		return errors.New("file is nil")
	}

	content, err := io.ReadAll(file)
	if err != nil {
		return err
	}

	values := strings.Fields(string(content))
	for i := len(values) - 1; i >= 0; i-- {
		s.Push(values[i])
	}
	return nil
}

func (s *Stack) SaveToBinaryFile(file *os.File) error {
	if file == nil {
		return errors.New("file is nil")
	}

	curr := s.top
	for curr != nil {
		strSize := uint64(len(curr.Data))
		err := binary.Write(file, binary.LittleEndian, strSize)
		if err != nil {
			return err
		}
		_, err = file.WriteString(curr.Data)
		if err != nil {
			return err
		}
		curr = curr.Next
	}
	return nil
}

func (s *Stack) LoadFromBinaryFile(file *os.File) error {
	if file == nil {
		return errors.New("file is nil")
	}

	var elements []string
	for {
		var strSize uint64
		err := binary.Read(file, binary.LittleEndian, &strSize)
		if err != nil {
			if err == io.EOF {
				break
			}
			return err
		}

		buffer := make([]byte, strSize)
		_, err = file.Read(buffer)
		if err != nil {
			return err
		}
		elements = append(elements, string(buffer))
	}

	for i := len(elements) - 1; i >= 0; i-- {
		s.Push(elements[i])
	}
	return nil
}